dttools/src/link.c

```diff
@@ -529,34 +518,31 @@ struct link *link_serve_address(const char *addr, int port)
 	return 0;
 }
 
-int link_ssl_wrap_server(struct link *link, const char *key, const char *cert) {
+int link_ssl_wrap_accept(struct link *link, const char *key, const char *cert) {
 #ifdef HAS_OPENSSL
 	if(key && cert) {
-		link->ctx = _create_ssl_context(/* is client */ 0);
-		_set_ssl_keys(link->ctx, key, cert);
+		debug(D_TCP, "accepting ssl state for %s port %d", link->raddr, link->rport);
 
-		return 1;
-	}
-#endif
-
-	return 0;
-}
-
-
-int link_ssl_wrap_accept(struct link *parent, struct link *link) {
-#ifdef HAS_OPENSSL
-	if(parent->ctx) {
-		debug(D_TCP, "setting up ssl state for %s port %d", link->raddr, link->rport);
+		/* go to blocking mode during handshake */
+		if(!link_nonblocking(link, 0)) {
+			return 0;
+		}
 
-		link->ctx = _create_ssl_context(/* is client */ 1);
+		link->ctx = _create_ssl_context();
+		_set_ssl_keys(link->ctx, key, cert);
 
-		link->ssl = SSL_new(parent->ctx);
+		link->ssl = SSL_new(link->ctx);
 		SSL_set_fd(link->ssl, link->fd);
 
 		int ret = SSL_accept(link->ssl);
 		if(ret <= 0) {
-			debug(D_SSL, "accept failed from %s port %d", link->raddr, link->rport);
+			debug(D_SSL, "ssl accept failed from %s port %d", link->raddr, link->rport);
 			ERR_print_errors_cb(_ssl_errors_cb, NULL);
+			ret = 0;
+		} else {
+			if(!link_nonblocking(link, 1)) {
```

Comment:
I think we want to return to non-blocking, regardless of the outcome of SSL_accept.